src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyMemoryLeakTest.java

```diff
@@ -49,9 +48,10 @@ public static void register(Object o) {
     @Test
     public void loaderReleased() throws Exception {
         FreeStyleProject p = r.jenkins.createProject(FreeStyleProject.class, "p");
-        String cp = CharacterPredicates.class.getProtectionDomain().getCodeSource().getLocation().toString(); // some JAR
-        p.getPublishersList().add(new TestGroovyRecorder(
-            new SecureGroovyScript(GroovyMemoryLeakTest.class.getName()+".register(this)", false, Collections.singletonList(new ClasspathEntry(cp)))));
+        String cp = GroovyMemoryLeakTest.class.getResource("somejar.jar").toString();
```

Comment:
So the problem with the test was that the previous class wasn't actually loaded from a separate jar?

Reply:
There were two problems with the previous revision of the test:

* The test did not actually load any classes from the specified classpath. So, it reproduced a leak when creating the outer cache, but not when adding entries to the inner cache. (Or perhaps dealt only with the resource cache, which does not suffer from the values-refers-to-key problem specific to the class cache fixed here, since a `URL` holds no interesting references.)
* The JAR file I used (`commons-text` or something) was in the test classpath (Maven `test` scope), so any attempts to load classes from it would actually go through `sun.misc.AppClassLoader` instead.

Reply:
> reproduced a leak when creating the outer cache, but not when adding entries to the inner cache. (Or perhaps dealt only with the resource cache…

Amendment: the class cache normally has a bunch of entries, including positives like `Object` or `WorkflowScript` as well as negatives for all sorts of Groovy nonsense. But in the previous test, none of the positive entries were for classes actually defined in the `URLClassLoader`—they were defined by lower loaders—so the fact that they were strongly held from the cache was irrelevant.

Reply:
To make it clearer still: when you actually load a class _from_ the classpath, there was until this fix a routine mistake of the sort that the `WeakHashMap` Javadoc warns about: a weakly held `URLClassLoader` key but then a strongly held `Class` value which in turn strongly holds the key. Pending my longstanding favorite [JDK-6389107](https://bugs.openjdk.java.net/browse/JDK-6389107), there is no easy universal fix for this problem. Nor is there any feasible way that I am aware of to mechanically detect the mistake at runtime (some profilers detect it for `WeakHashMap`, probably not for Guava caches); the `MemoryAssert` shown here can find it, if you think to write the right test!